### DISCUSSION TITLE
getting audio working

### DISCUSSION THREAD

**Discussion by cxrodgers** [wehr-lab/autopilot#55] [Q&A] [answered]
Baby steps.. Now working on getting hifiberry audio working. Run into a few areas of uncertainty. As always, apologies if I missed this somewhere in the doc!

- I ran the "jackd_apt" setup script. This triggered installation of a bunch of X11 stuff, because I did Raspbian Lite image which is lacking that. Just checking this is expected/normal? I'm unclear on why audio server would even require X11..
- When I start the pilot I get `Failed to connect to session bus for device reservation: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11` from jackd. Magical workaround is to `export DBUS_SESSION_BUS_ADDRESS=unix:path=/run/dbus/system_bus_socket` before calling autopilot. Or start an X server, I guess.
- Point of confusion: Is prefs['AUDIOSERVER'] supposed to be a bool or a string? This page suggests string https://docs.auto-pi-lot.com/en/latest/autopilot.stim.sound.sounds.html and this one suggests bool https://docs.auto-pi-lot.com/en/latest/_modules/autopilot/prefs.html?highlight=AUDIOSERVER. Probably related issue is "No Base Sound class specified! sounds will probably not work!!!", which I think happens if it's a bool.
- Next, when I start the pilot I get `Failed to acquire device name : Audio0 error : Connection ":1.27" is not allowed to own the service "org.freedesktop.ReserveDevice1.Audio0" due to security policies in the configuration file`. Something like this may help: https://www.raspberrypi.org/forums/viewtopic.php?t=57025

I'll keep working on this but probably the biggest open question I have is whether the Pilot is generally supposed to have a $DISPLAY attached or rather supposed to be headless? And confirming that jackd2 is the recommended audioserver? Thanks!!

**Comment by cxrodgers** [comment, marked as answer]
Actually this seems to work. I don't know why I didn't try it before as it is the workaround suggested in the jackd error message.
`export JACK_NO_AUDIO_RESERVATION=1`

**Reply by cxrodgers** [reply]
Yep, definitely working! Sheared off a few hair cells as well. Turns out volume 1.0 is pretty loud.

I think I can answer my own questions now:

- Install jackd2 and don't worry about the X11 dependencies that are installed
- `export JACK_NO_AUDIO_RESERVATION=1` before starting the Pilot
- set prefs['AUDIOSERVER'] to 'jack'
- Set amplitude to <0.1 and/or lower volume with `alsamixer` and/or build a hardware attenuator for the tweeters before testing :)

**Reply by sneakers-the-rat** [reply]
I'm going to accept this as the answer because it fixes the issue with using jackd on a headless pi without an X server, but with the note that i recommend building/installing from source if possible: https://github.com/wehr-lab/autopilot/discussions/55#discussioncomment-253677

**Comment by sneakers-the-rat** [comment]
There's a longstanding issue with jackd on a headless raspi, a newer version of jackd uses dbus which for seemingly multiple reasons (packed structs as mentioned here https://wiki.linuxaudio.org/wiki/raspberrypi#using_jack and the absence of an active X server), and the apt binaries build with dbus which makes them very finnicky and in my experience either completely fail or fail randomly.

I recommend building from source with the `jackd_source` script:
https://github.com/wehr-lab/autopilot/blob/606de487c73f7a807825028e5fee28d1e9815c3f/autopilot/setup/scripts.py#L105

but that also requires you to purge & remove all the installed files from the apt repository if you have installed it from there -- apt doesn't remove everything so it's a little messy but it essentially amounts to something like `rm /usr/bin/jack*` . 

This is also related to the last point in your question, i think building without dbus is in general a better idea, because not being able to reserve the device could lead to latency/lockup issues. Maybe this should be clearer in the scripts docs?

* Re the pref, that ambiguity is from having both the pyo and jackd audio servers implemented, but the jackd server is preferable for a number of reasons (not the least of which was frequent seemingly unavoidable dropout and long latencies from pyo), and so I have tacitly considered pyo deprecated for awhile, but have hesitated to remove it from the library altogether because it still could serve some use b/c it does provide a relatively mature way of doing live, parametric synthesis. Bear with me, still learning how this whole 'public development' process works. I think i'll put a deprecationwarning on pyo and make the pref a bool

The code can accept both strings and bools, setting it to `'jack'` is most explicit, but `True` behaves like `'jack'`. The somewhat clumsy import process that results in the warning has to do with the support for multiple base audio classes that have different dependencies (pyo used to be tricky to install, i think they've fixed that), while also needing to make the classes importable without the dependencies being present to build the docs. TMI but there you go.

sorry for the eardrums... need to add a speaker calibration routine for being able to set amplitude with dbSPL rather than some arbitrary scalar from 0-1....

This kinda hidden knowledge is a strong candidate for the type of stuff that i think should go on the wiki. I'm going to spend a bit of time now porting my notes/etc. from our lab wiki to the autopilot wiki and getting the basic structure of it started because it seems like that's a needed resource at this stage.

**Reply by cxrodgers** [reply]
Yes, this all makes great sense, thank you! I'm sure this info is present in the docs already and I just missed it. Writing good doc is like writing a paper, I think, hard to find the right linear order to transmit a large amount of non-linear information. I'll try to make a PR to clear up the parts that confused me.

> Bear with me, still learning how this whole 'public development' process works. 

Learning how to work better together is something we're all figuring out! Please know that the issues I post are not complaints but are for the sake of sharing information with other people who are getting to know the code better too. Also on that note, always feel free to ignore them whenever you are busy or around holidays, I post them more for historical record than for urgent help.

**Comment by sneakers-the-rat** [comment]
started to write this down here: https://wiki.auto-pi-lot.com/index.php/Jack_Audio
